### PR TITLE
F/introduce template options in rewrite rules

### LINKED ISSUE #652: rewrite rules don't honor time-zone
It seems to me this is a bug.
Ideally, one should be able to set `time-zone()` in a rewrite-rule.

#### config
```
@version:3.7
@include "scl.conf"

options {
  threaded(yes);
  time_zone("UTC");
};

source s_stdin { file('/dev/stdin'); };
rewrite r_utcdate { set("${ISODATE}", value('UTCDATE')); };
destination d_stdout { file('/dev/stdout' template ("ISODATE=${ISODATE} UTCDATE=${UTCDATE}\n")); };

log {
  source(s_stdin);
  rewrite(r_utcdate);
  destination(d_stdout);
};
```
#### Output
```
[2015-08-24T09:51:42.651988] syslog-ng starting up; version='3.7.1'
test
[2015-08-24T09:51:44.894269] Incoming log entry; line='test'
[2015-08-24T09:51:44.894394] Rewrite expression evaluation result; value='UTCDATE', new_value='2015-08-24T09:51:44+02:00', rule='r_utcdate', location='/etc/syslog-ng/syslog-ng-rewrite-timezone.conf:10:21'
[2015-08-24T09:51:44.894491] Initializing destination file writer; template='/dev/stdout', filename='/dev/stdout'
[2015-08-24T09:51:44.894800] Outgoing message; message='ISODATE=2015-08-24T07:51:44+00:00 UTCDATE=2015-08-24T09:51:44+02:00\x0a'
ISODATE=2015-08-24T07:51:44+00:00 UTCDATE=2015-08-24T09:51:44+02:00
```

### PR DESCRIPTION
**Pull request by ihrwein**
Fixes #652

### CODE REVIEW

File: lib/rewrite/rewrite-set.c [hunk 1]
```diff
@@ -70,10 +78,22 @@ log_rewrite_set_free(LogPipe *s)
...
   log_rewrite_free_method(s);
 }
 
+gboolean
+log_rewrite_set_init_method(LogPipe *s)
+{
+  LogRewriteSet *self = (LogRewriteSet *) s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+  log_rewrite_init_method(s);
+
```
Comment (bazsi) [inline lib/rewrite/rewrite-set.c]: Missing handling of the init method return value

File: lib/rewrite/rewrite-set.h [hunk 2]
```diff
@@ -29,5 +29,6 @@
 
 /* LogRewriteSet */
 LogRewrite *log_rewrite_set_new(LogTemplate *new_value, GlobalConfig *cfg);
+LogTemplateOptions *log_rewrite_set_get_value_template_options(LogRewrite *s);
```
Comment (bazsi) [inline lib/rewrite/rewrite-set.h]: I think we don't need the "value" term as qualification here. Makes all identifiers pretty long.

Just a nitpick.

File: lib/rewrite/rewrite-expr-grammar.ym [hunk 3]
```diff
@@ -158,9 +159,14 @@ rewrite_subst_opt
...
         : rewrite_expr_opt rewrite_expr_opts
+        | rewrite_set_opt rewrite_expr_opts
         |
         ;
 
+rewrite_set_opt
+        : template_option
+        ;
+
 rewrite_expr_opt
```
Comment (bazsi) [inline lib/rewrite/rewrite-expr-grammar.ym]: Usually these kind of grammars, we only use a single recursion and the _opt rule includes all potential options, set and generic expr alike.

File: lib/rewrite/rewrite-set.c [hunk 4]
```diff
@@ -43,7 +51,7 @@ log_rewrite_set_process(LogRewrite *s, LogMessage **pmsg, const LogPathOptions *
   GString *result;
 
   result = g_string_sized_new(64);
-  log_template_format(self->value_template, *pmsg, NULL, LTZ_LOCAL, 0, NULL, result);
+  log_template_format(self->value_template, *pmsg, &self->value_template_options, LTZ_SEND, 0, NULL, result);
```
Comment (bazsi) [inline lib/rewrite/rewrite-set.c]: I was thinking about the LTZ_SEND value. First I disagreed, but came to the conclusion that it's a good call. Set creates values that are being sent, and if no local setting is in place, the default should be send-time-zone().

So good call.

File: lib/rewrite/rewrite-expr-grammar.ym [hunk 5]
```diff
@@ -103,6 +103,7 @@ rewrite_expr
         | KW_SET '(' rewrite_template_content
           {
             last_rewrite = log_rewrite_set_new($3, configuration);
+            last_template_options = log_rewrite_set_get_value_template_options(last_rewrite);
             log_template_unref($3);
           }
```
Comment (bazsi) [inline lib/rewrite/rewrite-expr-grammar.ym]: What about the subst() rewrite op, don't we need the same there?

Comment (ihrwein) [inline lib/rewrite/rewrite-expr-grammar.ym]: I don't see the way to pass the `LogTemplateOptions` pointer trough `log_matcher_replace()`:

```
log_matcher_replace(self->matcher, *pmsg, self->super.value_handle, value, length, self->replacement, &new_length);
```

File: lib/rewrite/rewrite-expr-grammar.ym (hunk 3 above)
Comment (ihrwein) [inline lib/rewrite/rewrite-expr-grammar.ym]: This way `template_option` is only enabled to `rewrite-set`. Probably I can rephrase this to a simpler form.

**Comment by ihrwein** [comment]
@bazsi I fixed the review notes. This feature is not available in case of `subst` because I didn't find a way to pass around the `LogTemplateOptions*` in `log_matcher_replace()`:

```C
log_matcher_replace(LogMatcher *s, LogMessage *msg, gint value_handle, const gchar *value, gssize value_len, LogTemplate *replacement, gssize *new_length)
```

**Comment by bazsi** [comment]
hm, as an additional option? or, even better, it'd be nice to decouple
LogMatcher from templates by using a callback. Although I can see why this
would be a separate branch.


-- 
Bazsi

**Comment by ihrwein** [comment]
Do we have reason that `LogTemplate` doesn't contain a `LogTemplateOptions`? That way the grammar could use `log_template_get_options()` and set the time-zone and other options on it. Then `LogTemplateOptions` would be implicitly used by "everybody".

**Comment by bazsi** [comment]
I think the only reason is that they can be inherited from the global
settings. But that inheritance  could even be simpler if implemented in
LogTemplate.

This is probably a good idea, but one that should be explored first how it
would look like.

Templates are used in a lot of of contexts, and right now I don't know how
it would play out.

**Comment by dnsjts** [comment]
:+1:

**Comment by ihrwein** [comment]
@fekete-robert Could you update the docs? The `set` option in `rewrite` rules now supports the following well-known parameters:
* `ts-format`
* `frac-digits`
* `time-zone`
* `send-time-zone`
* `local-time-zone`
* `on-error`

**Comment by faxm0dem** [comment]
this is awesome thanks for all the good work!

**Comment by fekete-robert** [comment]
I've added to my todo list for OSE 3.8
